New feature/ids field finder

mgyoo86 opened the pull request:
## Summary
This PR introduces a new `findall` method that searches for specific fields within multiple `IDS` objects.
The target fields can be specified using either Symbols or Regex patterns.
The `findall` method returns an IDS_Field_Finder struct for each matched field, containing detailed information that allows for verbose terminal output and plot recipes.

Note: The plot `@recipes` will be represnted in another PR soon.

## Examples
Here are some quick examples.

### 1. Default behavior
```julia
# By default, it searches everything under the given IDS
# Following commands produce the same result
findall(dd.equilibrium.time_slice[].global_quantities)
findall(dd.equilibrium.time_slice[].global_quantities, r"") # Same behavior (Default)
findall(dd.equilibrium.time_slice[].global_quantities, :all) # Same behavior
```
<img src="https://github.com/user-attachments/assets/9fb762a4-9ca0-4ba9-875c-2da150d74c1c" width=80%>


### 2. Finding a single field within a single root IDS
```julia
 # Find fields matching a single symbol within a IDS structure
IFF = findall(dd.equilibrium.time_slice, :psi)
```
<img src="https://github.com/user-attachments/assets/4419bccc-89d0-4094-bc10-48544d54619e" width=80%>

### 3. Finding multiple fields within multiple root IDS objects
```julia
 # Search for multiple symbols within multiple root IDS objects
IFF = findall([dd.equilibrium, dd.core_sources], [:psi, :j_tor])
```
<img src="https://github.com/user-attachments/assets/135f080d-6e08-4daf-b25f-7159cd898b8f" width=80%>

### 4. Keyword argument of `include_subfields::Bool=true` 
```julia
  # Control subfield inclusion using the `include_subfields` keyword
IFF = findall(dd, r"prof.*2d"; include_subfields=false)
IFF = findall(dd, r"prof.*2d"; include_subfields=true) # Default behavior
```
<img src="https://github.com/user-attachments/assets/4415a1db-43c5-42a3-9dc5-0fdcfb24697e" width=80%>

### 5. Lazy evaluation of the value of found field objects

```julia
 # Retrieve actual values of found field objects (lazy evaluation)
IFF[1].value
IFF[end].value
```
<img src="https://github.com/user-attachments/assets/cc18f597-2151-4346-9fab-a588b6c2175b" width=50%>

orso82 commented on src/data.jl:
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
+    rest_part = replace(parent_name, root_name => "")
+
+    printstyled(io, root_name; color=:red)
+    if IFF.root_ids isa IMASdd.dd || root_name == parent_name
```
```suggestion
    if IFF.root_ids isa DD || root_name == parent_name
```
or
```suggestion
    if typeof(IFF.root_ids) <: DD || root_name == parent_name
```

orso82 commented on src/data.jl:
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
+
+    unit = units(IFF.parent_ids, IFF.field)
+    if !(isempty(unit) || unit == "-")
+        printstyled(io, " [$unit]"; color=:yellow, bold=true)
```
Yellow does not work on white background

orso82 commented on src/data.jl:
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
+    IFF_arr = Vector{IDS_Field_Finder}()
+    for root_ids in root_ids_arr
+        if root_ids isa Union{IDS,IDSvector}
+            append!(IFF_arr, Base.findall(root_ids, target_fields; include_subfields))
```
```suggestion
            append!(IFF_arr, findall(root_ids, target_fields; include_subfields))
```

orso82 commented on src/data.jl:
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
+end
+
+function Base.findall(root_ids::Union{IDS,IDSvector}, target::Union{Symbol,AbstractArray{Symbol},Regex}=r""; include_subfields::Bool=true)
+    target == :all ? target = r"" : target
```
```suggestion
```

orso82 commented on src/data.jl:
````diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
+# Example
+```julia-repl
+julia> findall(dd.equilibrium.time_slice[].global_quantities) # By default, it searches everything under given IDS
+julia> findall(dd.equilibrium.time_slice[].global_quantities, :all) # Same behavior
````
```suggestion
```

orso82 commented on src/data.jl:
```diff
@@ -867,6 +867,209 @@ function _match(@nospecialize(ids::IDSvector), conditions)
     return matches
 end
 
+struct IDS_Field_Finder
+    root_ids::Union{IDS,IDSvector} # Start point of the search
+    parent_ids::Union{IDS,IDSvector} # Parent IDS of target field
+    field::Symbol # Target field symbol
+    field_type::Type # Type of the field
+    field_path::String # Relative path from root_ids to the field
+
+    # Named constructor with keyword arguments for clear field assignment
+    IDS_Field_Finder(; root_ids, parent_ids, field, field_type, field_path) =
+        new(root_ids, parent_ids, field, field_type, field_path)
+end
```
```suggestion
Base.@kwdef struct IDS_Field_Finder
    root_ids::Union{IDS,IDSvector} # Start point of the search
    parent_ids::Union{IDS,IDSvector} # Parent IDS of target field
    field::Symbol # Target field symbol
    field_type::Type # Type of the field
    field_path::String # Relative path from root_ids to the field
end
```

orso82 reviewed (commented):
this is great! Thank you @mgyoo86 

I have some minor comments

orso82 commented:
I'd like to avoid this syntax. It doesn't add anything and it conflicts with the idea of plotting a quantity named `:all`
```
findall(dd.equilibrium.time_slice[].global_quantities, :all)
```

mgyoo86 commented:
Thanks for the quick and thorough review! @orso82 
I reflected on all of your comments as follows:
- changed color from ":yellow" to ":blue"
- removed the ":all" keywords.

orso82 commented:
I have tested it and merged it. It may be good to write a regression test for this.
